sys/linux/socket_ip_tunnel.txt

```diff
@@ -0,0 +1,93 @@
+# Copyright 2020 syzkaller project authors. All rights reserved.
+# Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+include <uapi/linux/if_tunnel.h>
+include <uapi/linux/if.h>
+include <uapi/linux/in.h>
+include <uapi/linux/ip.h>
+include <net/ip.h>
+
+resource sock_ipv4[sock]
+resource sock_ipv6[sock]
+
+socket$ipv4_tunnel(domain const[AF_INET], type const[SOCK_DGRAM], proto const[0]) sock_ipv4
+socket$ipv6_tunnel(domain const[AF_INET6], type const[SOCK_DGRAM], proto const[0]) sock_ipv6
+
+# Fallback tunnels and placeholder names "syztnl*".
+ipv4_tunnel_names = "", "ip_vti0", "tunl0", "gre0", "gretap0", "erspan0", "syztnl0", "syztnl1", "syztnl2"
+ipv6_tunnel_names = "", "ip6tnl0", "ip6_vti0", "ip6gre0", "sit0", "syztnl0", "syztnl1", "syztnl2"
+
+ip_tunnel_flags = GRE_CSUM, GRE_ROUTING, GRE_KEY, GRE_SEQ, GRE_STRICT, GRE_REC, GRE_ACK, GRE_FLAGS, GRE_VERSION, VTI_ISVTI
+ip_tunnel_protocols = 0, IPPROTO_IPIP, IPPROTO_GRE, IPPROTO_IPV6
+frag_off_flags = 0, IP_DF
+
+# Copied from vnet.txt:ipv4_header. `ipv4_header` requires the existence of the field
+# `ipv4_packet_t`. We only need the header here, so I copied the whole struct.
+iphdr {
+	ihl		bytesize4[parent, int8:4]
+	version		const[4, int8:4]
+	ecn		int8:2
+	dscp		int8:6
+	total_len	bytesize4[parent, int16be]
+	id		int16be[100:104]
+	frag_off	flags[frag_off_flags, int16be]
+	ttl		int8
+	protocol	flags[ip_tunnel_protocols, int8]
+	csum		csum[parent, inet, int16be]
+	src_ip		ipv4_addr
+	dst_ip		ipv4_addr
+} [packed]
```

Comment:
I'm not a syzkaller expert, but do we actually need to duplicate `iphdr` here?  Is there a way to "include" some other description file?

Reply:
[`ipv4_addr`](https://github.com/google/syzkaller/blob/c42af8f753155895a734373c820e76068251ba1b/sys/linux/vnet.txt#L419) has a member `total_len	len[ipv4_packet_t, int16be]`. However, in IP tunnels, only the header part is used, and if there is no `ipv4_packet_t`, there will be a compile error. An alternative solution might be adding a zero-sized `ipv4_packet_t`, but the last time I tried, it seems like syzkaller does not support zero-length buffer. Not sure what is the best way to do this?

Reply:
Looking closer, this iphdr looks wrong anyway.  I think the one you want is https://elixir.bootlin.com/linux/v5.8-rc3/source/include/uapi/linux/ip.h#L86.

```
$ grep -n 'iph\.' net/ipv4/ip_tunnel.c

97:             if (local != t->parms.iph.saddr ||
98:                 remote != t->parms.iph.daddr ||
112:            if (remote != t->parms.iph.daddr ||
113:                t->parms.iph.saddr != 0 ||
130:            if ((local != t->parms.iph.saddr || t->parms.iph.daddr != 0) &&
131:                (local != t->parms.iph.daddr || !ipv4_is_multicast(local)))
148:                t->parms.iph.saddr != 0 ||
149:                t->parms.iph.daddr != 0 ||
181:    if (parms->iph.daddr && !ipv4_is_multicast(parms->iph.daddr))
182:            remote = parms->iph.daddr;
213:    __be32 remote = parms->iph.daddr;
214:    __be32 local = parms->iph.saddr;
222:            if (local == t->parms.iph.saddr &&
223:                remote == t->parms.iph.daddr &&
524:            daddr = md ? dst : tunnel->parms.iph.daddr;
658:    connected = (tunnel->parms.iph.daddr != 0);
838:    t->parms.iph.saddr = p->iph.saddr;
839:    t->parms.iph.daddr = p->iph.daddr;
843:            memcpy(dev->dev_addr, &p->iph.saddr, 4);
844:            memcpy(dev->broadcast, &p->iph.daddr, 4);
848:    t->parms.iph.ttl = p->iph.ttl;
849:    t->parms.iph.tos = p->iph.tos;
850:    t->parms.iph.frag_off = p->iph.frag_off;
887:            if (p->iph.ttl)
888:                    p->iph.frag_off |= htons(IP_DF);
917:                            if (ipv4_is_multicast(p->iph.daddr))
919:                            else if (p->iph.daddr)
1204:                   if (ipv4_is_multicast(p->iph.daddr))
1206:                   else if (p->iph.daddr)
```

Reply:
Yes, you are right. Initially I used the definition of `iphdr` in the link you posted. However, there was a problem about the endianness (#1885), and after asking @dvyukov I decided to just copy the original `ipv4_addr`.
The two have the same structure and both follow the format of an IPv4 header, just that the field name are not the same.
I will change the names to match `iphdr` if it add more clarity. Thanks!

Reply:
Resolved by using `ipv4_packet_t[flags[ip_tunnel_protocols, int8], void]`.